Diffuse Roughness support

MiiBond opened the pull request:
This PR replaces https://github.com/BabylonJS/Babylon.js/pull/16183

This PR implements the base_diffuse_roughness parameter from the [OpenPBR specification](https://academysoftwarefoundation.github.io/OpenPBR/#model/basesubstrate).

The diffuse roughness is implemented for analytic lights, realtime-filtered IBL, prefiltered IBL and spherical harmonics.
I've added a flag to a material to choose between Lambert, Burley and the new Energy Conserving Oren-Nayar (EON) model used by OpenPBR. The default is OpenPBR's EON model. 

The previous behaviour was to use Burley diffuse for analytical lights and Lambert for IBL. Also, previously, specular roughness was applied to diffuse roughness for analytical lights while IBL didn't use it at all (because it was simply Lambertian). So, the new default slightly changes existing projects that used analytical lights but I question how noticeable that will be.

**Analytical Light:**
https://playground.babylonjs.com/?snapshot=refs/pull/16253/merge#MXACV7#3
![roughnessCompare](https://github.com/user-attachments/assets/df0f6782-f1c9-42b3-8c82-d2e4ebc59f85)

**Realtime IBL:**
https://playground.babylonjs.com/?snapshot=refs/pull/16253/merge#MXACV7#5
<img width="881" alt="image" src="https://github.com/user-attachments/assets/4d33b49c-a8d5-4903-85ca-78eb368c015e" />

The diffuse roughness models are heavily dependent on the light direction and view direction and are therefore difficult to handle with a prefiltered IBL. I came up with two methods for approximating roughness with prefiltered IBL's. The first, if we prefiltered using CDF, we generate a dominant light direction to use in the BRDF calculations. It works reasonably well.

**Prefiltered IBL with CDF:**
https://playground.babylonjs.com/?snapshot=refs/pull/16253/merge#MXACV7#9
<img width="883" alt="image" src="https://github.com/user-attachments/assets/b4c8383b-30be-4505-a09c-3bc5d6ee4cc3" />

The second approach, if you don't use CDF, is by approximating roughness by bending the surface normal towards the camera to add some of the retro-reflective behaviour that you get with EON. Because of this, you'll notice that Burley and EON are identical and the shadow terminator appears to move as diffuse roughness increases. In practice, however, this example uses an extreme IBL with a bright sunlight. With other IBL's, the effect tends to be more convincing.

**Prefiltered IBL without CDF**
https://playground.babylonjs.com/?snapshot=refs/pull/16253/merge#MXACV7#10
<img width="878" alt="image" src="https://github.com/user-attachments/assets/ffaa366c-6e9f-449a-8cbf-6364354e314f" />

The default IBL lighting in Sandbox uses spherical harmonics so we need to approximate diffuse roughness with this as well. I'm using the same bent normal technique as with prefiltered IBL without CDF.

**Spherical Harmonics IBL**
https://playground.babylonjs.com/?snapshot=refs/pull/16253/merge#MXACV7#11
![image](https://github.com/user-attachments/assets/13543a47-62aa-4d26-a438-1a6031803027)

bjsplat commented:
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

MiiBond commented on packages/dev/core/src/Materials/PBR/pbrBaseMaterial.ts:
```diff
@@ -2321,6 +2374,7 @@ export abstract class PBRBaseMaterial extends PushMaterial {
                 }
 
                 ubo.updateFloat("baseWeight", this._baseWeight);
+                ubo.updateFloat("baseDiffuseRoughness", this._baseDiffuseRoughness ?? this._roughness ?? 1);
```
I put this logic here to fallback on the old behaviour if baseDiffuseRoughness wasn't used. However, this was inconsistent between IBL and analytic lights so I suggest we just remove this old behaviour. What do you think?

MiiBond replied:
I've removed the fallback to the old behaviour now. How do you guys feel about this?
The roughness previously only had an effect on diffuse for analytical lights.

sebavan replied:
Would by default the rendering be impacted with the new code ? We had a few remarks/issues regarding the latest fixes in the PBR with ppl willing to use back the old behavior. If the impact is tiny it is ok but if it is more, we need at least a fallback flag

MiiBond replied:
Yes, there is a change to the default behaviour. It only affects rough surfaces that use analytical lights (the default behaviour of IBL is unaffected). Previously, the surface roughness was used as input to the Burley diffuse model but this was only ever used for analytical lights (i.e. direct lighting). IBL didn't use Burley and so diffuse irradiance from IBL was unaffected by roughness. This inconsistency is one of the reasons I think we should change the default behaviour.

Here's an example of the most extreme case where the change would be most noticeable. The sphere on the right has roughness 1.0.
Before the change:
<img width="679" alt="image" src="https://github.com/user-attachments/assets/627da837-3ff6-49e6-8b73-25962107697d" />

After the change:
<img width="691" alt="image" src="https://github.com/user-attachments/assets/d7561822-10b6-4a2d-9e40-3a186844d597" />

MiiBond replied:
Most cases won't be nearly this noticeable, of course. Diffuse roughness (for Burley or EON) is most noticeable when the view is aligned with the light direction, like in this example.

MiiBond replied:
Another point to note is that, if we change the default diffuse roughness model to EON, rather than Burley, the default look of these cases is going to change anyway.

I'm seeing that in a lot of these visualization test failures.

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

virtualzavie commented on packages/tools/tests/test/visualization/ReferenceImages/openpbr-base-diffuse-roughness-realtime-ibl.png:
Why is this render black? Does it not include your IBL implementation yet?

MiiBond replied:
Because I didn't check the output before I pushed my changes :)

Thanks for pointing this out. It's fixed now.

virtualzavie commented on packages/tools/tests/test/visualization/config.json:
```diff
@@ -6,6 +6,17 @@
+            "title": "OpenPBR Base Diffuse Roughness Realtime IBL",
+            "playgroundId": "#MXACV7#5",
+            "referenceImage": "openpbr_base_diffuse_roughness_realtime_ibl.png",
+            "excludedEngines": ["webgl1"]
```
Why is this excluded?

MiiBond replied:
Thanks for pointing that out. The CDF generator doesn't work in WebGL 1.0 but that should prevent this test from running. I've tested in WebGL 1.0 now and fixed a whole bunch of shader issues.
I think I can make the CDF generator work in WebGL 1.0 as well. I just have to get rid of all the texelFetch calls since that function doesn't exist in WebGL 1. Originally, the CDFGenerator was only for IBL Shadows which can't work in WebGL 1 because it doesn't support 3D textures.

virtualzavie commented on packages/dev/core/src/Shaders/ShadersInclude/hdrFilteringFunctions.fx:
```diff
@@ -242,7 +246,13 @@
+                            diffuseRoughnessTerm = vec3(diffuseBRDF_Burley(NoL, NoV, VoH, diffuseRoughness) * PI);
+                        #elif BASE_DIFFUSE_ROUGHNESS_MODEL == 2
+                            diffuseRoughnessTerm = diffuseBRDF_EON(vec3(1.0), diffuseRoughness, NoL, NoV) * PI;
+                        #endif
```
I'm a bit surprised there's a ×π factor for Burley and EON but not for Lambert. The result looks correct for Lambert though, so when does it happen? Is there a discrepency regarding when it's factored in?

MiiBond replied:
Both `diffuseBRDF_Burley` and `diffuseBRDF_EON` contain an explicit `1.0 / PI`.

MiiBond replied:
The pdf has a PI in it already, both for the CDF path and non-CDF path.
Burley and EON would double-up that computation so have an extra multiply by PI to cancel that out.

virtualzavie commented on packages/dev/core/src/Shaders/ShadersInclude/pbrBlockReflection.fx:
```diff
@@ -264,6 +266,7 @@
             #else
                 vec3 irradianceVector = vec3(reflectionMatrix * vec4(normalW, 0)).xyz;
             #endif
+            vec3 irradianceView = vec3(reflectionMatrix * vec4(viewDirectionW, 0)).xyz;
```
Wouldn't it be preferable to do:
```glsl
vec3 irradianceView = mat3(reflectionMatrix) * viewDirectionW;
```

virtualzavie commented on packages/dev/core/src/Shaders/ShadersInclude/pbrDirectLightingFunctions.fx:
```diff
@@ -41,7 +41,12 @@ vec3 computeHemisphericDiffuseLighting(preLightingInfo info, vec3 lightColor, ve
+        diffuseTerm = diffuseBRDF_Burley(info.NdotL, info.NdotV, info.VdotH, info.diffuseRoughness);
+    #elif BASE_DIFFUSE_ROUGHNESS_MODEL == 2
+        diffuseTerm = diffuseBRDF_EON(vec3(1.0), info.diffuseRoughness, info.NdotL, info.NdotV).x;
+    #endif
```
Why is the scale ×1/π ; ×1 ; ×1 here, whereas it was ×1 ; ×π ; ×π in the other implementation?

MiiBond replied:
This one is for direct lighting from an analytical light. The other was for the IBL filtering and I think the PI is already accounted for in the PDF.

virtualzavie replied:
I see. It makes sense, although that's not very consistent.

virtualzavie commented on packages/dev/core/src/Shaders/ShadersInclude/pbrDirectLightingFunctions.fx:
```diff
@@ -65,8 +70,12 @@ vec3 computeProjectionTextureDiffuseLighting(sampler2D projectionLightSampler, m
+            diffuseTerm = diffuseBRDF_Burley(NdotL, info.NdotV, info.VdotH, info.diffuseRoughness);
+        #elif BASE_DIFFUSE_ROUGHNESS_MODEL == 2
+            diffuseTerm = diffuseBRDF_EON(vec3(1.0), info.diffuseRoughness, NdotL, info.NdotV).x;
+        #endif
```
Same remark here.

virtualzavie commented on packages/dev/core/src/ShadersWGSL/ShadersInclude/pbrDirectLightingFunctions.fx:
```diff
@@ -35,7 +35,12 @@ fn computeHemisphericDiffuseLighting(info: preLightingInfo, lightColor: vec3f, g
+        diffuseTerm = diffuseBRDF_Burley(info.NdotL, info.NdotV, info.VdotH, info.diffuseRoughness);
+    #elif BASE_DIFFUSE_ROUGHNESS_MODEL == 2
+        diffuseTerm = diffuseBRDF_EON(vec3(1.0), info.diffuseRoughness, info.NdotL, info.NdotV).x;
+    #endif
```
Same remark here.

virtualzavie commented on packages/dev/core/src/ShadersWGSL/ShadersInclude/pbrDirectLightingFunctions.fx:
```diff
@@ -59,7 +64,12 @@ fn computeProjectionTextureDiffuseLighting(projectionLightTexture: texture_2d<f3
+            diffuseTerm = diffuseBRDF_Burley(NdotL, info.NdotV, info.VdotH, info.diffuseRoughness);
+        #elif BASE_DIFFUSE_ROUGHNESS_MODEL == 2
+            diffuseTerm = diffuseBRDF_EON(vec3(1.0), info.diffuseRoughness, NdotL, info.NdotV).x;
+        #endif
```
And here.

virtualzavie reviewed (commented):
Continued review.

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

MiiBond commented on packages/tools/tests/test/visualization/ReferenceImages/openpbr-base-diffuse-roughness.png:
Yeah, I thought the same thing but it seems as though this retro-reflectivity is expected when viewing directly along the direction of the light. When viewing from the side, it gets a lot darker.
<img width="341" alt="image" src="https://github.com/user-attachments/assets/ab0c2fdb-56ee-4cf6-853e-596c2929870d" />
However, it is still slightly brighter than it should be and this is due to me post-multiplying the albedo rather than passing the albedo into the EON computation. For analytical lights and realtime filtering of the IBL, we can do EON correctly but I also want this to work for prefiltered IBL's so I want a solution where we can post-multiply the albedo. I'm working on improving this...

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

bjsplat commented:
Snapshot stored with reference name:
refs/pull/16253/merge

Test environment: 
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/16253/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/16253/merge
https://gui.babylonjs.com/?snapshot=refs/pull/16253/merge
https://nme.babylonjs.com/?snapshot=refs/pull/16253/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/16253/merge#BCU1XR#0

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

Popov72 commented on packages/dev/core/src/Shaders/ShadersInclude/pbrBlockReflection.fx:
```diff
@@ -297,7 +305,29 @@
+                float NoV = dot(irradianceVector, irradianceView);
+                
+                vec3 diffuseRoughnessTerm = vec3(1.0);
+                #if BASE_DIFFUSE_ROUGHNESS_MODEL == 0
+                    float LoV = dot (Ls, irradianceView);
+                    float mag = length(reflectionDominantDirection) * 2.0;
+                    diffuseRoughnessTerm = diffuseBRDF_EON(vec3(1.0), diffuseRoughness, NoL, NoV, LoV) * PI;
+                    diffuseRoughnessTerm = mix(vec3(1.0), diffuseRoughnessTerm, sqrt(min(mag * NoV, 1.0)));
+                #elif BASE_DIFFUSE_ROUGHNESS_MODEL == 1
+                    vec3 H = (irradianceView + Ls)*0.5;
+                    float VoH = dot(irradianceView, H);
+                    diffuseRoughnessTerm = vec3(diffuseBRDF_Burley(NoL, NoV, VoH, diffuseRoughness) * PI);
+                #endif
```
Do you have any pointers that would help us understand these calculations?

In the [Frostbite doc (page 70)](https://seblagarde.wordpress.com/wp-content/uploads/2015/07/course_notes_moving_frostbite_to_pbr_v32.pdf), they calculate a dominant direction, but it's used to query the irradiance map, not to alter the result of the query.

MiiBond replied:
So, diffuse roughness depends on both the viewing angle and light direction. Here, I'm trying to modify the existing Lambertian prefiltered IBL to approximate EON and Burley using a single light direction for the IBL.
Basically, it has the effect of brightening glancing angles when the light and view vector align.
The magnitude of the dominant direction is kind of a measure of how strongly anisotropic the lighting in the IBL is.

Popov72 commented on packages/dev/core/src/Shaders/ShadersInclude/pbrBlockReflection.fx:
```diff
@@ -264,7 +267,12 @@
+            vec3 irradianceView = vec3(reflectionMatrix * vec4(viewDirectionW, 0)).xyz;
+            #if !defined(USE_IRRADIANCE_DOMINANT_DIRECTION) && !defined(REALTIME_FILTERING)
+                float NdotV = max(dot(normalW, viewDirectionW), 0.0);
+                irradianceVector = mix(irradianceVector, irradianceView, (0.5 * (1.0 - NdotV)) * diffuseRoughness);
```
Doesn't this calculation depend on the type of the diffuse BRDF we are currently using? In other words, will Burley have a different calculation from EON (Lambert is view independent, so we shouldn't execute this code in that case)?

Also, this is a breaking change in Burley case (our default diffuse), I don't know if we should do something about it? cc @sebavan

MiiBond replied:
So, this is the bent normal hack to approximate diffuse roughness for IBL's if you aren't doing realtime filtering or prefiltering with CDF. This works for SH as well and I don't have a different version for Burley or EON. They use the same logic.

Yes, I suppose we could disable this if the user has chosen Lambert as the diffuse roughness mode. That's basically the same as disabling diffuse roughness.

Popov72 commented on packages/dev/core/src/Shaders/ShadersInclude/pbrBlockReflectivity.fx:
```diff
@@ -32,6 +33,11 @@ reflectivityOutParams reflectivityBlock(
+    , in float baseDiffuseRoughness
+#ifdef BASE_DIFFUSE_ROUGHNESS
+    , in float baseDiffuseRoughnessTexture
+    , in vec2 vBaseDiffuseRoughnessInfos
```
```suggestion
    , in vec2 baseDiffuseRoughnessInfos
```
to avoid potential mismatch with `vBaseDiffuseRoughnessInfos`, which is the uniform passed to the shader.

Popov72 commented on packages/dev/core/src/Shaders/ShadersInclude/pbrBlockReflectivity.fx:
```diff
@@ -160,8 +166,14 @@ reflectivityOutParams reflectivityBlock(
 
+    float diffuseRoughness = baseDiffuseRoughness;
+#ifdef BASE_DIFFUSE_ROUGHNESS
+    diffuseRoughness *= baseDiffuseRoughnessTexture;
```
```suggestion
    diffuseRoughness *= baseDiffuseRoughnessTexture * baseDiffuseRoughnessInfos.y;
```
`baseDiffuseRoughnessInfos.y` is the global intensity of the texture.

Popov72 commented on packages/dev/core/src/Shaders/ShadersInclude/pbrFragmentDeclaration.fx:
```diff
@@ -67,6 +72,9 @@ uniform mat4 view;
     #endif
     uniform mat4 reflectionMatrix;
     uniform vec3 vReflectionMicrosurfaceInfos;
+    #ifdef USEIRRADIANCEMAP
```
```suggestion
    #if defined(USEIRRADIANCEMAP) && defined(USE_IRRADIANCE_DOMINANT_DIRECTION)
```

Popov72 commented on packages/dev/core/src/Shaders/iblDominantDirection.fragment.fx:
```diff
@@ -0,0 +1,27 @@
+precision highp sampler2D;
+precision highp samplerCube;
+
+#include<helperFunctions>
+#include<importanceSampling>
+#include<pbrBRDFFunctions>
+#include<hdrFilteringFunctions>
```
I think you only need the `hdrFilteringFunctions` include.

MiiBond replied:
If any one of these is missing, it doesn't compile.

Popov72 replied:
My bad, the functions from `hdrFilteringFunctions` that we don't use still need these includes...

Popov72 commented on packages/dev/core/src/Shaders/hdrIrradianceFiltering.fragment.fx:
```diff
@@ -13,7 +13,7 @@ uniform float hdrScale;
 
 void main() {
-    vec3 color = irradiance(inputTexture, direction, vFilteringInfo
+    vec3 color = irradiance(inputTexture, direction, vFilteringInfo, 1.0, direction
```
I think you only need the `hdrFilteringFunctions` include.

Popov72 commented on packages/dev/core/src/Shaders/pbr.vertex.fx:
```diff
@@ -175,7 +176,11 @@ void main(void) {
+        vec3 viewDirectionW = normalize(vEyePosition.xyz - vPositionW);
+        float NdotV = max(dot(vNormalW, viewDirectionW), 0.0);
+        // Bend the normal towards the viewer based on the diffuse roughness
+        vec3 roughNormal = mix(vNormalW, viewDirectionW, (0.5 * (1.0 - NdotV)) * baseDiffuseRoughness);
```
Same question as above regarding the calculation for Burley / EON and the breaking change.

MiiBond replied:
I can disable this in the case that `BASE_DIFFUSE_ROUGHNESS_MODEL` is Lambert.

Popov72 commented on packages/dev/inspector/src/components/actionTabs/tabs/propertyGrids/materials/pbrMaterialPropertyGridComponent.tsx:
```diff
@@ -353,6 +377,16 @@ export class PBRMaterialPropertyGridComponent extends React.Component<IPBRMateri
                         step={0.01}
                         onPropertyChangedObservable={this.props.onPropertyChangedObservable}
                     />
+                    <SliderLineComponent
+                        lockObject={this.props.lockObject}
+                        label="Base Diffuse Roughness"
+                        target={material}
+                        propertyName="baseDiffuseRoughness"
+                        minimum={0}
+                        maximum={1}
+                        step={0.01}
+                        onPropertyChangedObservable={this.props.onPropertyChangedObservable}
+                    />
```
We shouldn't have this slider in two different sections "Lighting & Colors" and "Metallic Workflow", it could be confusing for the user. Maybe keep it only in the latter, as "Roughness" is to be found in that section?

MiiBond replied:
Yeah, I'm not sure how I managed that... :)

Popov72 reviewed (changes requested):
Great job!

Regarding the spherical harmonic case, I think our calculation will have to be updated to take into account the new EON model?

We will have to update our .env generation tool to also generate the irradiance map for EON. cc @sebavan.

cc @alexchuber for the glTF serializer part, and @bghgary for the glTF loader part.

[EDIT] Also be patient, as several members of the team are on leave this week.

MiiBond commented on packages/dev/core/src/Engines/constants.ts:
```diff
@@ -447,6 +447,10 @@ export class Constants {
      */
     public static readonly MATERIAL_CounterClockWiseSideOrientation = 1;
 
+    public static readonly MATERIAL_DIFFUSE_ROUGHNESS_OPENPBR = 0;
```
Yes, I think that's a good idea.

MiiBond commented on packages/dev/core/src/Materials/PBR/pbrBaseMaterial.ts:
```diff
@@ -804,6 +820,18 @@ export abstract class PBRBaseMaterial extends PushMaterial {
         this.markAsDirty(Constants.MATERIAL_TextureDirtyFlag);
     }
 
+    private _baseDiffuseRoughnessModel: number = Constants.MATERIAL_DIFFUSE_ROUGHNESS_OPENPBR;
```
No matter what the default model is, it's still a breaking change because the previous behaviour already wasn't consistent. Burley was being used for analytical lights while Lambert was always used for IBL.
We could make Burley the default to help with this but the behaviour has still changed. Also, previously, the specular roughness was being applied to the Burley diffuse.

bghgary commented on packages/dev/loaders/src/glTF/2.0/Extensions/EXT_materials_diffuse_roughness.ts:
```diff
@@ -0,0 +1,102 @@
+    }
+}
+
+/**
+ * [Specification](https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Khronos/EXT_materials_diffuse_roughness/README.md)
+ * [Playground Sample](https://www.babylonjs-playground.com/frame.html#7F7PN6#8)
+ */
```
Link is dead. Playground sample is for something else. Spec is PR, so we should mark as experimental.
```suggestion
/**
 * [Specification](https://github.com/KhronosGroup/glTF/blob/fdee35425ae560ea378092e38977216d63a094ec/extensions/2.0/Vendor/EXT_materials_diffuse_roughness/README.md)
 * @experimental
 */

```

MiiBond commented on packages/dev/core/src/Materials/PBR/pbrMaterial.ts:
```diff
@@ -284,6 +291,13 @@ export class PBRMaterial extends PBRBaseMaterial {
+    @serialize("baseDiffuseRoughness")
+    @expandToProperty("_markAllSubMeshesAsTexturesDirty")
+    public baseDiffuseRoughness = 0;
+
```
I think I don't understand the relationship between pbrBaseMaterial and pbrMaterial as far as these properties are concerned. Could I please get a brief explanation about how I should be adding material properties?

Popov72 replied:
`PBRBaseMaterial` is our (internal) PBR implementation, it cannot be used by the end-user (the class is abstract).

It has been done this way so that we can build classes over it, like `PBRMaterial` (gives access to the full PBR implementation), `PBRMetallicRoughnessMaterial` (gives only access to the metallic workflow) and `PBRSpecularGlossinessMaterial` (gives only access to the specular/glossiness workflow). The latter two classes exist mostly for historical purposes, I think that everyone now uses `PBRMaterial` exclusively.

So, when adding new features, you must update `PBRBaseMaterial` because it is the PBR implementation, and `PBRMaterial` to let end-users have access to these new features.

MiiBond replied:
I guess I still don't understand what should be in `pbrBaseMaterial` and what should be in `pbrMaterial`.
Some things, like `realTimeFilteringQuality` have private and public get/set in only in `pbrBaseMaterial`. Other properties, like `_albedoColor` are private in `pbrBaseMaterial` and have a public version without the underscore in `pbrMaterial`. But I don't see how these are even connected to each other. How does `albedoColor` get the value from `_albedoColor`? I don't see anywhere in code where this is happening.

MiiBond replied:
@Popov72

MiiBond commented on packages/dev/core/src/Rendering/iblCdfGenerator.ts:
```diff
@@ -369,6 +397,46 @@ export class IblCdfGenerator {
+        return new Promise((resolve) => {
+            this._dominantDirectionPT.onGeneratedObservable.addOnce(() => {
+                const data = new Float32Array(4);
+                this._dominantDirectionPT.readPixels(0, 0, data, true)?.then((value) => {
```
Typescript complains that `this._dominantDirectionPT.readPixels(0, 0, data, true)` could be `null`

Popov72 replied:
Indeed. You can use “!” instead of “?”, because we know that the buffer cannot be null, since the returned buffer is the one we pass as the 3rd parameter to `readPixels`.

MiiBond commented on packages/dev/core/src/Materials/Textures/Filtering/hdrIrradianceFiltering.ts:
```diff
@@ -217,6 +217,12 @@ export class HDRIrradianceFiltering {
+            await this._cdfGenerator.findDominantDirection().then((dir) => {
+                irradianceTexture.dominantDirection = dir;
+            });
+        }
```
The irradiance prefilter doesn't care about the diffuse BRDF with the current implementation. Diffuse roughness is variable (using a texture) and I didn't want to generate multiple prefiltered IBL's.
1. For realtime filtering, this isn't an issue and we can do everything per-pixel.
2. For prefiltered IBL with CDF, I generate the dominant direction and use that to modify the intensity of the IBL based on NdotV and diffuse roughness. The IBL is just the standard Lambert irradiance though.
3. For prefiltered IBL without CDF, I do a hack to bend the normal based on the diffuse roughness to try to mimic the effect of a diffuse rough surface. Still using the Lambertian irradiance.
4. For SH, I do the same bent-normal hack.

Popov72 replied:
That's fine for me, cc @sebavan to be sure we don't want to generate specific IBLs for each types of BRDF and always use the Lambert irradiance map.

MiiBond commented on packages/dev/core/src/Shaders/ShadersInclude/pbrBRDFFunctions.fx:
```diff
@@ -424,6 +424,45 @@ float diffuseBRDF_Burley(float NdotL, float NdotV, float VdotH, float roughness)
+
+vec3 diffuseBRDF_EON(vec3 albedo, float roughness, float NdotL, float NdotV, float LdotV)
+{
+    // Because surface albedo is applied later in the shader, it is awkward to
+    // include it here only for EON. Instead, we use an implicit albedo of 0.5
+    // and then divide by it at the end.
```
No, it's not quite the same result. It's a pretty close approximation though.

MiiBond commented on packages/public/glTF2Interface/babylon.glTF2Interface.d.ts:
```diff
@@ -1183,6 +1186,15 @@ declare module BABYLON.GLTF2 {
+     * Interfaces from the EXT_materials_diffuse_roughness extension
+     */
+    /** @internal */
+    interface IKHRMaterialsDiffuseRoughness {
```
Good catch! Thanks!

MiiBond commented:
Thanks for all the comments, guys. I think I've addressed most of them but we probably need to have some discussions about backwards compatibility?
I'm not sure there is a perfect solution since we had the inconsistency before between Lambert for IBL and Burley for analytical lights.
The case where things change is if the user had a rough model with analytical lights. Once getting this change, the material will be specularly rough but will have a Lambertian diffuse (diffuse_roughness = 0) instead of rough Burley. I don't think it's likely to be a big change.

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

Popov72 commented on packages/dev/core/src/Shaders/ShadersInclude/hdrFilteringFunctions.fx:
```diff
@@ -214,20 +217,36 @@
                     vec3 Ls = uv_to_normal(vec2(1.0 - fract(T.x + 0.25), T.y));
                     float NoL = dot(n, Ls);
+                    float NoV = dot(n, inputV);
+                    #if BASE_DIFFUSE_ROUGHNESS_MODEL == 0 // EON
```
Let me have a look at it.

Popov72 replied:
This PR should make it work:

https://github.com/BabylonJS/Babylon.js/pull/16474

MiiBond replied:
Thank you. Works like a charm!

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

Popov72 commented:
Yes, let's discuss backward compatibility with @sebavan when he gets back.

To simplify matters, I have reproduced here the various comments on backward compatibility:

-----
P: Do we really want the OpenPBR model to be the default model? It's a breaking change, and because it's not a bug fix, it's a hard change to justify...
MB: No matter what the default model is, it's still a breaking change because the previous behaviour already wasn't consistent. Burley was being used for analytical lights while Lambert was always used for IBL.

[Not generating multiple prefiltered IBLs]
That's fine with me, cc @sebavan to make sure we don't want to generate specific IBLs for each type of BRDF and always use the Lambert irradiance map.

[`dominantDirection` property added to `BaseTexture`]
This property is really specific to the irradiance map, I don't think it's right to put it in `BaseTexture`...
Maybe we need a `IBLTexture` class, but I don't really know how/where it would fit in the current framework if we don't want any breaking changes...

sebavan commented:
@MiiBond considering that by default diffuse roughness would be equal to specular roughness (our current roughness) The main  change is moving from Burley to Oren Nayar s diffuse model.

Could you generate three pictures (roughness 0, 0.5 and 1) with increasing metalness for increasing gray scales spheres ?

This could help understand the extends of the breaking change ? in the mean time could you revert to our current model by default ? so the next PR could be only a simple switch flip ?

sebavan commented:
[Not generating multiple prefiltered IBLs]
All good, lambert is fine there

[dominantDirection property added to BaseTexture]
We unfortunately need to store it there :-( but yup I wonder how/where we should move all the env specific properties like rotation low mid and high texture for lack of texture lods, harmonics and so on

Popov72 commented:
Regarding the dominant direction, I think we should tag it @internal (and rename it `_dominantDirection`), because there's no reason a end-user would use it, and we would be free to refactor the code in the future without incurring a breaking change.

MiiBond commented:
@sebavan I'm not sure if this is what you're asking for but here's a quick comparison of the only case where backwards compatibility will be lost. The only change is that `roughness` will no longer affect **diffuse** roughness for analytical lights. It only affects specular roughness now because we have a separate parameter for the diffuse roughness.

Here's what it used to look like for rough spheres with a single directional light:
https://playground.babylonjs.com/?snapshot=refs/pull/16253/merge#MXACV7#17

And here's what it will look like now for rough spheres with a single directional light:
https://playground.babylonjs.com/?snapshot=refs/pull/16253/merge#MXACV7#18

MiiBond commented:
@sebavan I've changed the default to Burley and also changed the logic so that if `diffuseRoughness` isn't set, we'll use the value from `roughness` instead. Note that this will change the look of IBL irradiance on rough surfaces, compared to the current behaviour.

@Popov72 I've marked `_dominantDirection` @internal.

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

MiiBond commented:
Note that a lot more of the visualization tests have now failed because we're defaulting the diffuse roughness to the specular roughness when not set. This was the previous behaviour but now we support diffuse roughness for IBL's so any Playground that uses rough materials and IBL will look different.

sebavan commented:
Oh is it not equivalent to the legacy code if we do diffuse roughness = spec roughness in burley mode ?

Cause I guess in this case we should more aim at back compat in default case ?

Popov72 commented on packages/dev/core/src/Shaders/ShadersInclude/pbrBRDFFunctions.fx:
```diff
@@ -429,6 +429,45 @@ float diffuseBRDF_Burley(float NdotL, float NdotV, float VdotH, float roughness)
+    vec3 f_ms = (rho_ms * RECIPROCAL_PI) * max(eps, 1.0 - EFo) // multi-scatter lobe
+                                 * max(eps, 1.0 - EFi)
+                                 / max(eps, 1.0 - avgEF);
+    return (f_ss + f_ms) / rho;
```
What about using the right albedo value for the calculation (and not 0.5), and return `(f_ss + f_ms) / rho / albedo`? This way, when we multiply the diffuse contribution by albedo later on, we will end up with the right value for diffuse.

MiiBond replied:
Done

MiiBond commented:
> Cause I guess in this case we should more aim at back compat in default case ?

I think the most backwards-compatible behaviour would be to default `baseDiffuseRoughness` to 0.0 so that diffuse lighting is effectively Lambertian for every type of lighting. This effects only the backwards compatibility of rough surfaces for analytical lights.

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

Popov72 commented:
@MiiBond I don't know if you missed them, but just in case, there are still a few issues to be resolved, which are hidden under the link "48 hidden items":

![image](https://github.com/user-attachments/assets/1fd1ed67-e06e-4461-a2e4-6d9f8be88949)

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

virtualzavie reviewed (commented):
Very partial review as I'm reading this implementation.

MiiBond commented on packages/dev/core/src/Engines/constants.ts:
```diff
@@ -447,6 +447,21 @@ export class Constants {
+    /**
+     * Lambertian diffuse model type.
+     */
+    public static readonly MATERIAL_DIFFUSE_ROUGHNESS_LAMBERT = 2;
```
True. Maybe `MATERIAL_DIFFUSE_ROUGHNESS_NONE` is better?

MiiBond commented on packages/dev/core/src/Materials/Node/Blocks/PBR/pbrMetallicRoughnessBlock.ts:
```diff
@@ -270,6 +271,19 @@ export class PBRMetallicRoughnessBlock extends NodeMaterialBlock {
     })
     public realTimeFilteringQuality = Constants.TEXTURE_FILTERING_QUALITY_LOW;
 
+    /**
+     * Base Diffuse Roughness Model
+     */
+    @editableInPropertyPage("Diffuse Roughness Model", PropertyTypeForEdition.List, "RENDERING", {
```
Ah, yeah, that's a good point. I like that more.

MiiBond commented on packages/tools/tests/test/visualization/ReferenceImages/openpbr-base-diffuse-roughness-prefiltered-ibl.png:
Yup. This scene is prefiltering without CDF and the IBL has a strong sunlight in it so I expect noise. I could have chosen a different IBL but I thought I'd keep it consistent with the other test scenes.

virtualzavie replied:
Ah ok, without a CDF that's expected indeed.

MiiBond commented on packages/tools/tests/test/visualization/ReferenceImages/openpbr-base-diffuse-roughness-spherical-harmonics.png:
This is just what the spherical harmonics look like with this IBL. I assume it's just the lack of accuracy that you get with SH but I couldn't say for sure that it's correct.
I didn't change the SH generation or anything though so, if there is an issue, I don't think we should worry about it in this PR.

virtualzavie replied:
Ok, I'll let @sebavan look if this is expected or if something's amiss.
I was just surprised to see this render being much darker than the others.

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

MiiBond commented:
> there are still a few issues to be resolved

@Popov72 Thank you, I did miss some of those. I've addressed them all now, I think, except for the `pbrBaseMaterial` VS `pbrMaterial` question.

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

sebavan commented:
> > Cause I guess in this case we should more aim at back compat in default case ?
> 
> I think the most backwards-compatible behaviour would be to default `baseDiffuseRoughness` to 0.0 so that diffuse lighting is effectively Lambertian for every type of lighting. This effects only the backwards compatibility of rough surfaces for analytical lights.

Is there a way for full back compat ? In case this breaks prod apps ?

MiiBond commented:
> Is there a way for full back compat ? In case this breaks prod apps ?

Well, we could hard code the analytical lighting functions to always use Burley with specular roughness and ignore diffuse roughness completely.
Do you have a communications channel for breaking changes?

sebavan commented:
You can ask on the forum, as we had quite a few users coming back to use for broken back compat on previous changes and we had to add back compat flags :-( When the impact is mid/large visually we need some solutions for it even more when it is not a bug fix.

Popov72 commented on packages/dev/core/src/Materials/PBR/pbrBaseMaterial.ts:
```diff
@@ -2232,6 +2299,18 @@ export abstract class PBRBaseMaterial extends PushMaterial {
+                                }
+                            } else {
+                                ubo.updateVector3("vReflectionDominantDirection", Vector3.Up());
+                            }
```
I think you can safely simplify with the code I suggested?

MiiBond replied:
Ah, thanks.
Done.

MiiBond replied:
@Popov72

MiiBond commented:
I'd like to suggest that we don't add a special flag to enable the legacy behaviour. Here is my reasoning:
1. The legacy behaviour was inconsistent in that it applied Burley diffuse only for analytical lights and not for IBL. This makes adding a backwards-compatibility flag pretty awkward and messy.
2. The change is not noticeable in most cases and pretty subtle when it is. Existing projects will only appear different if they use analytical lights and rough surfaces. Only the diffuse lobe changes and only when viewing the surface from the direction of the light. For example, here is the most extreme comparison between Burley and Lambert. A fully-rough sphere, viewed directly along the directional light vector. When viewed from the side, there is no difference.
![Roughness1](https://github.com/user-attachments/assets/bca0706d-9431-46e4-97b4-c4aea5b0dea1)
![Roughness1-side](https://github.com/user-attachments/assets/18d926aa-e99a-4965-a661-89bf5eb3e28e)

4. The legacy appearance can be achieved by choosing Burley as the diffuse model and setting diffuse roughness to match specular roughness.

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

sebavan commented:
The point 3. is selling it to me. As long as we have a Documented way to go back to the previous renders, I am totally fine with it.

Now which model should be the default is the tricky one ?

MiiBond commented:
Besides, the EON model is energy conserving and looks better, imho.

MiiBond commented:
Are there release notes that we can modify with the relevant info?

MiiBond commented:
Do I update that now or after this is merged?
If before, is that the only thing preventing this from being merged?

Popov72 commented:
You should update the documentation after the PR has been merged and a new version has been published, so that you can indicate this version number in the doc.

I think we are ok to merge: cc @sebavan

MiiBond commented:
@sebavan I've added the static member for the default diffuse model.

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

Popov72 commented:
@MiiBond I think there's still this comment left:

![image](https://github.com/user-attachments/assets/e3d14a9d-5ab2-478c-aec5-fbaeec745dec)

bjsplat commented:
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

MiiBond commented:
@Popov72 @sebavan 
The linter in the CI is giving me this error but I don't get it locally:
```
error @typescript-eslint/naming-convention: Interface name `GLTFLoaderExtensionOptions` must have one of the following prefixes: I
```

I'll make this change but I wanted to call this out.

bjsplat commented:
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

sebavan commented:
> @Popov72 @sebavan The linter in the CI is giving me this error but I don't get it locally:
> 
> ```
> error @typescript-eslint/naming-convention: Interface name `GLTFLoaderExtensionOptions` must have one of the following prefixes: I
> ```
> 
> I'll make this change but I wanted to call this out.

I do not think we should change it

cc @RaananW

RaananW commented:
> @Popov72 @sebavan The linter in the CI is giving me this error but I don't get it locally:
> 
> ```
> error @typescript-eslint/naming-convention: Interface name `GLTFLoaderExtensionOptions` must have one of the following prefixes: I
> ```
> 
> I'll make this change but I wanted to call this out.

No need to replace the entire interface, you will need to add `eslint-disable-next-line @typescript-eslint/naming-convention` before the line using it (where you used it).

Fetch from master and run `npm install` to be sure you are up to date with the newest eslint rules

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16253/merge/testResults/webgl2playwright/index.html
